packages/ERTP/src/amountMath.js

```diff
@@ -54,138 +53,118 @@ export { MathKind };
+  if (brand !== undefined) {
+    mustBeComparable(brand);
+    assert.equal(amount.brand, brand);
+  }
+};
 
-    // Return the amount included in leftAmount but not included in
-    // rightAmount. If leftAmount does not include all of rightAmount,
-    // error.
-    subtract: (leftAmount, rightAmount) =>
-      amountMath.make(
-        helpers.doSubtract(
-          amountMath.getValue(leftAmount),
-          amountMath.getValue(rightAmount),
-        ),
-      ),
-  });
-  const empty = amountMath.make(helpers.doGetEmpty());
-  return amountMath;
-}
+const noCoerceMake = (value, brand) => {
+  const amount = harden({ brand, value });
+  return amount;
+};
 
-harden(makeAmountMath);
+/** @type {AmountMath} */
+const amountMath = {
+  make: (allegedValue, brand) => {
+    mustBeComparable(brand);
+    const value = sniffHelpers(allegedValue).doCoerce(allegedValue);
+    const amount = harden({ brand, value });
+    return amount;
+  },
+  coerce: (allegedAmount, brand) => {
+    mustBeComparable(brand);
+    const { brand: allegedBrand, value } = allegedAmount;
+    assert(
+      allegedBrand !== undefined,
+      details`The brand in allegedAmount ${allegedAmount} is undefined. Did you pass a value rather than an amount?`,
+    );
+    assert(
+      brand === allegedBrand,
+      details`The brand in the allegedAmount ${allegedAmount} in 'coerce' didn't match the amountMath brand ${brand}.`,
+    );
+    // Will throw on inappropriate value
+    return amountMath.make(value, brand);
+  },
+  getValue: (amount, brand) => {
+    mustBeComparable(brand);
+    return amountMath.coerce(amount, brand).value;
+  },
+  makeEmpty: mathKind => {
+    assert(
+      helpers[mathKind],
+      details`${mathKind} must be MathKind.NAT, MathKind.SET, or MathKind.STRING_SET`,
+    );
+    return helpers[mathKind].doGetEmpty();
+  },
+  isEmpty: (amount, brand = undefined) => {
+    optionalBrandCheck(amount, brand);
+    return sniffHelpers(amount).doIsEmpty(amount.value);
+  },
+  isGTE: (leftAmount, rightAmount, brand = undefined) => {
+    optionalBrandCheck(leftAmount, brand);
+    optionalBrandCheck(rightAmount, brand);
+    assert.equal(leftAmount.brand, rightAmount.brand);
+    return sniffHelpers(leftAmount).doIsGTE(
+      leftAmount.value,
+      rightAmount.value,
+    );
+  },
+  isEqual: (leftAmount, rightAmount, brand = undefined) => {
+    optionalBrandCheck(leftAmount, brand);
+    optionalBrandCheck(rightAmount, brand);
+    assert.equal(leftAmount.brand, rightAmount.brand);
+    return sniffHelpers(leftAmount).doIsEqual(
+      leftAmount.value,
+      rightAmount.value,
+    );
+  },
+  add: (leftAmount, rightAmount, brand = undefined) => {
+    optionalBrandCheck(leftAmount, brand);
+    optionalBrandCheck(rightAmount, brand);
+    assert.equal(leftAmount.brand, rightAmount.brand);
+    return noCoerceMake(
+      sniffHelpers(leftAmount).doAdd(leftAmount.value, rightAmount.value),
+      leftAmount.brand,
+    );
+  },
```

Comment:
I'll just use `add` as an example of the issue that occurs across all the binary operators. Since you look up the helper only from the left operand, even if `sniffHelpers` was accurate, this would try to add `harden(['foo'])` and `harden([{}])` together with `strSetMathHelpers`, which would not work.